ENH: Add MGC to scipy.stats

sampan501 opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://docs.scipy.org/doc/numpy/dev/gitwash/development_workflow.html#writing-the-commit-message

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->
None

#### What does this implement/fix?
<!--Please explain your changes.-->
This PR adds MGC, a high dimensional independence test that is powerful enough to extract linear and nonlinear trends. `scipy.stats` contains a host of independence and other hypothesis tests, but are limited by assumptions of normality, linearity, unidimensionality, etc. While this may be appropriate in a host of circumstances, it is increasingly important to analyze nonlinear and high dimensional trends. The tests included can also be used to enable both classification and also for regression.

#### Additional information
<!--Any additional information you think is important.-->
I sent previous message board posts in June here: [https://mail.python.org/pipermail/scipy-dev/2019-June/023544.html](https://mail.python.org/pipermail/scipy-dev/2019-June/023544.html) and here: [https://mail.python.org/pipermail/scipy-dev/2019-June/023544.html](https://mail.python.org/pipermail/scipy-dev/2019-June/023566.html).

In addition, information about `mgcpy` the package from which MGC comes from is here: [https://github.com/neurodata/mgcpy](https://github.com/neurodata/mgcpy) and the paper describing it is here: [https://arxiv.org/abs/1907.02088](https://arxiv.org/abs/1907.02088).

tylerjereddy commented on scipy/stats/_stats.pyx:
```diff
@@ -300,3 +301,165 @@ def _weightedrankedtau(ordered[:] x, ordered[:] y, intp_t[:] rank, weigher, bool
+@cython.wraparound(False)
+@cython.boundscheck(False)
+cpdef _local_covariance(np.ndarray[np.float_t, ndim=2] distx,
+                        np.ndarray[np.float_t, ndim=2] disty,
```
are we tightly bound to this older buffer syntax instead of modern typed memoryviews? I know I've had a hard time transitioning some of our cython to the newer format, but for new code it is nice to try!

sampan501 replied:
I think so. I think that it's fine to use `memoryviews` for two of the more basic functions, but for many of them, I am doing a lot of matrix arithmetic operations with `ndarrays`. Converting between the two becomes inefficient and actually raises run time by about 10 seconds for unit tests. For larger arrays, this may correspond to hours longer. If you know any workarounds for this I would be happy to implement it!

peterbell10 replied:
>Converting between the two becomes inefficient and actually raises run time by about 10 seconds for unit tests.

In case this isn't obvious, you should never have to convert from view->`ndarray`  since manipulating the memory view modifies the array's contents inplace. e.g.
```python
float [:,:] cov_xy_view = cov_xy

for k in range(nx - 1):
    for l in range(ny - 1):
        cov_xy_view[k+1, l+1] += (
            cov_xy_view[k+1, l] + cov_xy_view[k, l+1]
            - cov_xy_view[k, l])
   
# Continue using cov_xy as normal
cov_xy[0, :] = np.cumsum(cov_xy[0, :])
```

Using the memory views for function arguments (e.g. in `_dense_rank_data`) is not a good idea since you now have to two additional constructors `ndarray`->view->`ndarray` where before there were none.

sampan501 replied:
https://github.com/scipy/scipy/pull/10524/commits/01fe019ff13f4b1664e9d0b34886d997224f9652

tylerjereddy commented on scipy/stats/_stats.pyx:
```diff
@@ -300,3 +301,165 @@ def _weightedrankedtau(ordered[:] x, ordered[:] y, intp_t[:] rank, weigher, bool
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+cpdef _dense_rank_data(np.ndarray[np.float_t, ndim=1] x):
```
here and in a few other places `cpdef` functions are defined, but it seems the functions are only used in Cython--perhaps there is some room for simplification by only using `cdef` when calls from Python aren't used / envisioned?

sampan501 replied:
https://github.com/scipy/scipy/pull/10524/commits/a9e9b1f5a7673163a2c1d046c72c2c692a212212

tylerjereddy commented on scipy/stats/stats.py:
```diff
@@ -4226,6 +4232,395 @@ def weightedtau(x, y, rank=True, weigher=None, additive=True):
+
+    Examples
+    --------
+    >>> import numpy as np
```
I think we can safely assume this import has already happened for our refguide_check infrastructure

rgommers replied:
indeed, this line should be left out of all examples

sampan501 replied:
https://github.com/scipy/scipy/pull/10524/commits/a97997aa164c6b6f2d737e41e4e3a4f59a9ad9ba

tylerjereddy commented on scipy/stats/stats.py:
```diff
@@ -4226,6 +4232,395 @@ def weightedtau(x, y, rank=True, weigher=None, additive=True):
+
+
+def _euclidean_dist(x):
+    return squareform(pdist(x, metric='euclidean'))
```
Isn't this `cdist(x, x)`? If defining a private function to calculate Euclidean distance is really justified, maybe a brief comment is in order, since we have so many ways of doing this already

sampan501 replied:
The reason that `squareform(pdist(x, metric='euclidean'))` was used was so that pairwise distances are always going to be calculated for this input. If I was to use `cdist(x, x)` as the default, that would require `compute_distance` to be of the form `compute_distance(x, x)` rather than just `compute_distance(x)`.

I could probably just change `squareform(pdist(x, metric='euclidean'))` to `cdist(x, x)` if that works?

sampan501 replied:
https://github.com/scipy/scipy/pull/10524/commits/390e52a9eab7e4c29c0dea72cd0814fe26445fac

tylerjereddy commented on scipy/stats/tests/test_stats.py:
```diff
@@ -5079,3 +5084,138 @@ def test_names(self):
+
+        # raises error if reps is None type
+        reps = ()
+        assert_raises(ValueError, stats.mgc, x, x, reps=reps)
```
parametrizing over `reps` might work nicely here

sampan501 replied:
https://github.com/scipy/scipy/pull/10524/commits/02c1c5d98aff32492837bc3bb69305706e581a2a

rgommers commented on doc/source/tutorial/stats.rst:
```diff
@@ -1070,3 +1070,87 @@ the individual data points on top.
+
+The simulation relationship can be plotted below:
+
+.. plot:: tutorial/stats/plots/mgc_plot2.png
```
It would be better to check in the Python script that generates these images. There are 4 existing examples like that in `doc/source/tutorial/stats/plots/`

sampan501 replied:
https://github.com/scipy/scipy/pull/10524/commits/cf60cdb53d14cda8c6d73e39fb6a92c87b153ab9

rgommers commented:
Just did a brief scan so far, but this looks like it's in quite good shape, thanks @sampan501.

I think in the mailing list discussion we didn't talk about the function name yet. `mgc` is fairly non-descriptive. "multiscale graph correlation" is long, but perhaps something like `multiscale_graphcorr` would be a decent name?

sampan501 commented:
> Just did a brief scan so far, but this looks like it's in quite good shape, thanks @sampan501.
> 
> I think in the mailing list discussion we didn't talk about the function name yet. `mgc` is fairly non-descriptive. "multiscale graph correlation" is long, but perhaps something like `multiscale_graphcorr` would be a decent name?

https://github.com/scipy/scipy/pull/10524/commits/17560a13b9e80912b27f269686ac73d212c88e3f

sampan501 commented:
Hello all,

I believe I have addressed all the feedback so far. I believe that everything is working correctly on my end except for the tutorial which asserts correctly when I have been testing locally and renders correctly online. Also, the scipy master build seems to be failing and not sure what to do about that.

rgommers commented:
> the scipy master build seems to be failing and not sure what to do about that.

test failures are mostly due to this PR it looks like (although there's an unrelated one for `hilbert`).

This is a TravisCI failure:
```
File "doc/source/tutorial/stats.rst", line 1126, in /home/travis/build/scipy/scipy/tools/../doc/source/tutorial/stats.rst
Failed example:
    compute_mgc(x, y)
Expected:
    MGC test statistic: 0.7
    P-value: 0.0
    Optimal scale: [20, 20]
Got:
    MGC test statistic:  0.7
    P-value:  0.0
    Optimal scale:  [21, 21]
File "doc/source/tutorial/stats.rst", line 1154, in /home/travis/build/scipy/scipy/tools/../doc/source/tutorial/stats.rst
Failed example:
    compute_mgc(x, y)
Expected:
    MGC test statistic: -0.1
    P-value: 0.9
    Optimal scale: [20, 20]
Got:
    MGC test statistic:  -0.1
    P-value:  0.9
    Optimal scale:  [21, 21]
```

This is on the Azure builds, both Windows and 32-bit Linux:
```
>           raise ValueError("Number of reps must be an integer greater than 0.")
E           ValueError: Number of reps must be an integer greater than 0.

compute_distance = <function _euclidean_dist at 0xebc772b4>
reps       = 1000
x          = array([[ 0.09762701],
       [ 0.43037873],
       [ 0.20552675],
       [ 0.08976637],
       [-0.1526904 ],
       [...353343],
       [ 0.34127574],
       [-0.57923488],
       [-0.7421474 ],
       [-0.3691433 ],
       [-0.27257846]])
y          = array([[ 0.09762701],
       [ 0.43037873],
       [ 0.20552675],
       [ 0.08976637],
       [-0.1526904 ],
       [...353343],
       [ 0.34127574],
       [-0.57923488],
       [-0.7421474 ],
       [-0.3691433 ],
       [-0.27257846]])

scipy/stats/stats.py:4456: ValueError
```

sampan501 commented:
> test failures are mostly due to this PR it looks like (although there's an unrelated one for `hilbert`).

Looking through everything now, I can fairly confidently say that everything is integrated correctly. It's not clear to me if the `build_docs` is due to my code or someone else's (but I think probably not mine since Travis CI builds my docs correctly).

sampan501 commented:
Ok, I believe that everything has built correctly! If you have any more feedback let me know

peterbell10 commented on scipy/stats/_stats.pyx:
```diff
@@ -300,3 +301,170 @@ def _weightedrankedtau(ordered[:] x, ordered[:] y, intp_t[:] rank, weigher, bool
+@cython.boundscheck(False)
+cdef _center_distance_matrix(ndarray distx, str global_corr='mgc', is_ranked=True):
+    cdef int n = distx.shape[0]
+    cdef ndarray rank_distx = np.zeros((<object> distx).shape)
```
Does this cast do anything?

sampan501 replied:
I changed it to make this bit of code less confusing https://github.com/scipy/scipy/pull/10524/commits/fcfe4ea8a0c3eedfb76d7911ae4d95e986d2dda8

peterbell10 commented on scipy/stats/_stats.pyx:
```diff
@@ -300,3 +301,170 @@ def _weightedrankedtau(ordered[:] x, ordered[:] y, intp_t[:] rank, weigher, bool
+    # convert float32 numpy array to int, as it will be used as array indices
+    # [0 to n-1]
+    rank_distx = rank_distx.astype(np.int) - 1
+    rank_disty = rank_disty.astype(np.int) - 1
```
This should come before the memory view initialisation. It might also be better to use `asarray` or `copy=False` to avoid copies.

peterbell10 commented on scipy/stats/_stats.pyx:
```diff
@@ -300,3 +301,170 @@ def _weightedrankedtau(ordered[:] x, ordered[:] y, intp_t[:] rank, weigher, bool
+    # summing up the the element-wise product of A and B based on the ranks,
+    # yields the local family of covariances
+    cdef float a, b
+    cdef int i, j, k, l
```
```suggestion
    cdef intp_t i, j, k, l
```

sampan501 commented:
If there is no more feedback, I think that everything builds correctly minus some issues in unit tests that are not mine.

rgommers commented:
Implementation wise this is looking quite good, thanks @sampan501.

It would be awesome if someone could try this on an existing real-world example they have and see if the functionality and API "feel right".

I won't have time this week, but will try to play with it after that.

j1c commented:
@rgommers I just tried the `multiscale_graphcorr` and it works great. I would say the API is as expected from the docstrings provided. See the minimal code below. You will need the [graspy](https://github.com/neurodata/graspy) package to run the following code.

```import numpy as np
from scipy.stats import multiscale_graphcorr

from graspy.embed import AdjacencySpectralEmbed
from graspy.datasets import load_drosophila_left, load_drosophila_right

left = load_drosophila_left()
right = load_drosophila_right()

left_embeddings = np.hstack(AdjacencySpectralEmbed(3).fit_transform(left))
right_emebeddings = np.hstack(AdjacencySpectralEmbed(3).fit_transform(right))

stacked_embeddings = np.vstack([left_embeddings, right_emebeddings])
labels = np.array([0] * left_embeddings.shape[0] + [1] * right_emebeddings.shape[0]).reshape(-1, 1)

res = multiscale_graphcorr(stacked_embeddings, labels)

print(f"{res.stat:.6f}")
0.055124

print(res.pvalue)
0.001
```

sampan501 commented:
@rgommers Have you had a chance to try out the functionality and API?

rgommers commented on scipy/stats/tests/test_stats.py:
```diff
@@ -5079,3 +5084,144 @@ def test_names(self):
+                      compute_distance=compute_distance)
+
+    @pytest.mark.parametrize("reps", [
+        -1,    # reps is negative
```
looks like this should be `'-1'` (a string)?

sampan501 replied:
Here I am verifying that when `reps` is negative as specified in this line of the source code:

`if not isinstance(reps, int) or reps < 0:`

It should raise a value error

rgommers replied:
ah okay, I didn't get the comment - a mix of string and ints is usually a mistake (there `'1'` as a string on the next line)

sampan501 commented on doc/source/tutorial/stats.rst:
```diff
@@ -1069,3 +1069,170 @@ the individual data points on top.
+    ...         mgc_map = mgc_dict["mgc_map"]
+    ...         ax = plt.gca()
+    ...         # draw heatmap
+    ...         plt.title("Local Correlation Map", fontsize=17)
```
https://github.com/scipy/scipy/pull/10524/commits/16ab7e18e05305c08f3b31adcdc9c212380d6068

sampan501 commented on doc/source/tutorial/stats.rst:
```diff
@@ -1069,3 +1069,170 @@ the individual data points on top.
+
+Let's look at some linear data first:
+
+    >>> x = np.array([-0.69198779, 0.18236784, -0.55349325, -0.29817661,
```
https://github.com/scipy/scipy/pull/10524/commits/16ab7e18e05305c08f3b31adcdc9c212380d6068

sampan501 commented on doc/source/tutorial/stats.rst:
```diff
@@ -1069,3 +1069,170 @@ the individual data points on top.
+The same can be done for nonlinear data sets. The following :math:`x` and
+:math:`y` arrays are derived from a nonlinear simulation:
+
+    >>> x = np.array([-0.57765724, -2.9276214, -1.04262652, 1.2582997,
```
https://github.com/scipy/scipy/pull/10524/commits/16ab7e18e05305c08f3b31adcdc9c212380d6068

sampan501 commented on doc/source/tutorial/stats.rst:
```diff
@@ -1069,3 +1069,170 @@ the individual data points on top.
+   :include-source: 0
+
+Now, we can see the test statistic, p-value, and MGC map visualized below. The
+optimal scale is shown on the map as a red "x":
```
https://github.com/scipy/scipy/pull/10524/commits/16ab7e18e05305c08f3b31adcdc9c212380d6068

sampan501 commented on doc/source/tutorial/stats/plots/mgc_plot1.py:
```diff
@@ -0,0 +1,47 @@
+    plt.show()
+
+
+x = np.array([-0.69198779, 0.18236784, -0.55349325, -0.29817661, -0.18634447,
```
https://github.com/scipy/scipy/pull/10524/commits/16ab7e18e05305c08f3b31adcdc9c212380d6068

sampan501 commented on scipy/stats/stats.py:
```diff
@@ -4226,6 +4232,391 @@ def weightedtau(x, y, rank=True, weigher=None, additive=True):
+    Examples
+    --------
+    >>> from scipy.stats import multiscale_graphcorr
+    >>> x = np.array([0.07487683, -0.18073412, 0.37266440, 0.06074847,
```
https://github.com/scipy/scipy/pull/10524/commits/4afd6ec4071e105d2a4854d5a1684dbeeae40beb

sampan501 commented on scipy/stats/stats.py:
```diff
@@ -4226,6 +4232,391 @@ def weightedtau(x, y, rank=True, weigher=None, additive=True):
+    statistic.
+
+    MGC requires at least 5 samples to run with reliable results. It can also
+    handle multidimensional arrays.
```
https://github.com/scipy/scipy/pull/10524/commits/4afd6ec4071e105d2a4854d5a1684dbeeae40beb

sampan501 commented on scipy/stats/stats.py:
```diff
@@ -4226,6 +4232,391 @@ def weightedtau(x, y, rank=True, weigher=None, additive=True):
+    Returns
+    -------
+    stat : float
+        The sample MGC test statistic within :math:`[-1, 1]`.
```
https://github.com/scipy/scipy/pull/10524/commits/4afd6ec4071e105d2a4854d5a1684dbeeae40beb

sampan501 commented on scipy/stats/stats.py:
```diff
@@ -4226,6 +4232,391 @@ def weightedtau(x, y, rank=True, weigher=None, additive=True):
+
+    Building upon the ideas of nearest neighbor methods and distance-based
+    statistical methods, MGC achieves higher statistical power compared with
+    comparable distance based statistics while also providing a map that
```
https://github.com/scipy/scipy/pull/10524/commits/4afd6ec4071e105d2a4854d5a1684dbeeae40beb

jovo replied:
@rgommers thanks for all your help/comments.  i fully agree.  it took me like a year to write that paragraph clearly, i fully support using it verbatim :) that snippet you included does not explicitly include the term "latent geometry" , so i think it is satisfactory to you as is?

rgommers replied:
awesome, if you as the original author agree, then we can include it verbatim (I was worrried about copyright, but I checked at the bottom of the article that it's yours) - even better. thanks @jovo

sampan501 replied:
https://github.com/scipy/scipy/pull/10524/commits/2eb4608a50207fe42c3d1a54c2924d94403eb8f2

sampan501 commented on scipy/stats/_stats.pyx:
```diff
@@ -300,3 +301,168 @@ def _weightedrankedtau(ordered[:] x, ordered[:] y, intp_t[:] rank, weigher, bool
+            cov_xy_view[k+1, l+1] += (cov_xy_view[k+1, l] + cov_xy_view[k, l+1] - cov_xy_view[k, l])
+
+    # centering the covariances
+    cov_xy = cov_xy - ((expectx.reshape(-1, 1) @ expecty.reshape(-1, 1).T) / n**2)  # caveat when porting from R (reshape)
```
https://github.com/scipy/scipy/pull/10524/commits/4afd6ec4071e105d2a4854d5a1684dbeeae40beb

sampan501 commented on scipy/stats/_stats.pyx:
```diff
@@ -300,3 +301,168 @@ def _weightedrankedtau(ordered[:] x, ordered[:] y, intp_t[:] rank, weigher, bool
+    # 2 caveats when porting from R (np.sqrt and reshape)
+    corr_mat = cov_mat / np.sqrt(local_varx.reshape(-1, 1) @ local_vary.reshape(-1, 1).T).real
+    # avoid computational issues that may cause a few local correlations
+    # to be negligebly larger than 1
```
https://github.com/scipy/scipy/pull/10524/commits/4afd6ec4071e105d2a4854d5a1684dbeeae40beb

sampan501 commented on scipy/stats/_stats.pyx:
```diff
@@ -300,3 +301,168 @@ def _weightedrankedtau(ordered[:] x, ordered[:] y, intp_t[:] rank, weigher, bool
+
+    # normalizing the covariances yields the local family of correlations
+
+    warnings.filterwarnings("ignore")
```
https://github.com/scipy/scipy/pull/10524/commits/4afd6ec4071e105d2a4854d5a1684dbeeae40beb

sampan501 commented on scipy/stats/stats.py:
```diff
@@ -4226,6 +4232,391 @@ def weightedtau(x, y, rank=True, weigher=None, additive=True):
+        distance.
+    reps : int, optional
+        The number of replications used to estimate the null when using the
+        permutation test. The default is 1000 repelications.
```
https://github.com/scipy/scipy/pull/10524/commits/4afd6ec4071e105d2a4854d5a1684dbeeae40beb

sampan501 commented on scipy/stats/stats.py:
```diff
@@ -4226,6 +4232,391 @@ def weightedtau(x, y, rank=True, weigher=None, additive=True):
+
+def multiscale_graphcorr(x, y, compute_distance=_euclidean_dist, reps=1000):
+    r"""
+    Computes the Multiscale Graph Correlation (MGC) test statistic, a high
```
https://github.com/scipy/scipy/pull/10524/commits/4afd6ec4071e105d2a4854d5a1684dbeeae40beb

jovo commented:
yup, its mine, and i fully agree! thanks!

On Wed, Aug 14, 2019 at 3:16 PM Ralf Gommers <notifications@github.com>
wrote:

> *@rgommers* commented on this pull request.
> ------------------------------
>
> In scipy/stats/stats.py
> <https://github.com/scipy/scipy/pull/10524#discussion_r314040138>:
>
> > +
> +def _euclidean_dist(x):
> +    return cdist(x, x)
> +
> +
> +MGCResult = namedtuple('MGCResult', ('stat', 'pvalue', 'mgc_dict'))
> +
> +
> +def multiscale_graphcorr(x, y, compute_distance=_euclidean_dist, reps=1000):
> +    r"""
> +    Computes the Multiscale Graph Correlation (MGC) test statistic, a high
> +    dimensional measure of independence between arbitrary data matrices.
> +
> +    Building upon the ideas of nearest neighbor methods and distance-based
> +    statistical methods, MGC achieves higher statistical power compared with
> +    comparable distance based statistics while also providing a map that
>
> awesome, if you as the original author agree, then we can include it
> verbatim (I was worrried about copyright, but I checked at the bottom of
> the article that it's yours) - even better. thanks @jovo
> <https://github.com/jovo>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scipy/scipy/pull/10524?email_source=notifications&email_token=AAAKG4QI7NDF523TNRGWIYDQERKVLA5CNFSM4IG7B732YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCBSZ6GA#discussion_r314040138>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAKG4VLRW7YPCEIOD4FP7LQERKVLANCNFSM4IG7B73Q>
> .
>


-- 
the glass is all full:  half water, half air.
(i will be on paternity leave for most of september and october)
neurodata.io

sampan501 commented:
@rgommers I know that in the discussion boards we were talking about the advantages and disadvantages of OpenMP. Is it worth implementing that in `multiscale_graphcorr`?

rgommers commented:
> @rgommers I know that in the discussion boards we were talking about the advantages and disadvantages of OpenMP. Is it worth implementing that in `multiscale_graphcorr`?

OpenMP not, but parallelization yes. Assuming that that makes sense for this function. You wouldn't call this in an inner loop, but if calculation time becomes prohibitive for large input sizes, then adding a `workers` keyword and using `multiprocessing` or some thread pool could be a good idea. 

Would you want to add parallelization given the current performance?

sampan501 commented:
> Would you want to add parallelization given the current performance?

I think so, just because the performance is limited by the speed of the permutation test, which calculates the test statistic 1000 times to approximate the null distribution. Maybe parallelizing this loop will improve performance?

rgommers commented:
> Maybe parallelizing this loop will improve performance?

It should indeed. You can copy the approach of `optimize.differential_evolution`, using `workers=1`

sampan501 commented:
@rgommers Have you had a chance to look at the algorithm in-depth?

sampan501 commented:
Has there been anything that needs to be changed in the algorithm as currently constructed?

rgommers commented:
Sorry for the delay @sampan501. This is on my todo list, but that list was only growing the last couple of weeks. I'll try to review again on the weekend.

loftusa commented:
@rgommers any updates?

rgommers commented:
> @rgommers any updates?

My todo list is only getting longer it seems. I did have another look through and am tempted to merge as is. Everything looks good to me; not verifying in every last detail that algorithm matches the paper is probably forgivable.

@andyfaff could you have a quick look at the `MapWrapper` and `_ParallelP` code in this PR?

rgommers commented on scipy/stats/stats.py:
```diff
@@ -4259,6 +4265,529 @@ def weightedtau(x, y, rank=True, weigher=None, additive=True):
+    if y.ndim == 1:
+        y.shape = (-1, 1)
+
+    nx, px = x.shape
```
This will raise `ValueError: too many values to unpack (expected 2)` if the input wasn't 1-D or 2-D. Perhaps insert an explicit check above and raise a more informative `ValueError`?
```
if x.ndim == 1:
    ...
elif x.ndim != 2:
    raise ValueError("Expected a 2-D array `x1`, found shape {}".format(x.shape))
```

sampan501 replied:
https://github.com/scipy/scipy/pull/10524/commits/fabfbb13356c806099a19d3ce7ed5448ba2219d0

rgommers commented on scipy/stats/stats.py:
```diff
@@ -4259,6 +4265,529 @@ def weightedtau(x, y, rank=True, weigher=None, additive=True):
+                         "results.")
+
+    # convert x and y to float
+    x = np.asarray(x).astype(np.float64)
```
you already checked that `x` and `y` were `ndarray` instances, so `asarray` can be removed

sampan501 replied:
https://github.com/scipy/scipy/pull/10524/commits/fabfbb13356c806099a19d3ce7ed5448ba2219d0

rgommers commented on scipy/stats/stats.py:
```diff
@@ -4259,6 +4265,529 @@ def weightedtau(x, y, rank=True, weigher=None, additive=True):
+
+    n, m = stat_mgc_map.shape
+    if m == 1 or n == 1:
+        stat = stat_mgc_map[m - 1][n - 1]
```
this deserves a comment. why is `stat` the last element of the array if `stat_mgc_map` is a 1-D array?

sampan501 replied:
https://github.com/scipy/scipy/pull/10524/commits/fabfbb13356c806099a19d3ce7ed5448ba2219d0

rgommers commented on scipy/stats/stats.py:
```diff
@@ -4259,6 +4265,529 @@ def weightedtau(x, y, rank=True, weigher=None, additive=True):
+    """
+    m, n = stat_mgc_map.shape
+
+    # parametric threshold, based on a threshold is estimated based on the
```
this sentence needs a rephrase.

also why the 0.02 value? give a reference if there's a justification somewhere (I assume there is)

sampan501 replied:
https://github.com/scipy/scipy/pull/10524/commits/fabfbb13356c806099a19d3ce7ed5448ba2219d0

rgommers commented on scipy/stats/stats.py:
```diff
@@ -4259,6 +4265,529 @@ def weightedtau(x, y, rank=True, weigher=None, additive=True):
+
+    if np.linalg.norm(sig_connect) != 0:
+        # proceed only when the connected region's area is sufficiently large
+        if np.sum(sig_connect) >= np.ceil(0.02 * max(m, n)) * min(m, n):
```
another 0.02 value that deserves a comment/reference

sampan501 replied:
https://github.com/scipy/scipy/pull/10524/commits/fabfbb13356c806099a19d3ce7ed5448ba2219d0

rgommers commented on scipy/stats/stats.py:
```diff
@@ -4259,6 +4265,529 @@ def weightedtau(x, y, rank=True, weigher=None, additive=True):
+        1 refers to samples that come from `v`. `y` thus has shape `(2n, 1)`.
+    """
+    x = np.concatenate([u, v], axis=0)
+    y = np.concatenate([np.repeat(0, u.shape[0]), np.repeat(1, v.shape[0])],
```
`np.repeat(0, ...)` --> `np.zeros(...)`

sampan501 replied:
https://github.com/scipy/scipy/pull/10524/commits/fabfbb13356c806099a19d3ce7ed5448ba2219d0

rgommers commented on scipy/stats/tests/test_stats.py:
```diff
@@ -5079,3 +5084,147 @@ def test_names(self):
+        y = np.arange(3)
+        assert_raises(ValueError, stats.multiscale_graphcorr, x, y)
+
+    def test_error_nans(self):
```
Could you add a test for arrays with `inf`s in them? It's not handled or filtered out, so what happens when `inf`s propagate and is that the expected result?

sampan501 replied:
https://github.com/scipy/scipy/pull/10524/commits/47e4cb6b00cc6558b10e8986ad759d55e0be01e5

jovo commented:
@rgommers wow, thanks so much, all such useful feedback, definitely will markedly improve functionality and readability and usability.  @sampan501 will address it all asap!

sampan501 commented on doc/release/1.4.0-notes.rst:
```diff
@@ -69,6 +69,10 @@ lambda expressions for them.
 
 `scipy.stats` improvements
 --------------------------
+:func:`scipy.stats.multiscale_graphcorr` added as an independence test that
```
https://github.com/scipy/scipy/pull/10524/commits/2582d2c72aaef8735e0c7868a4b4de0e5f5c05b3

sampan501 commented on scipy/stats/_stats.pyx:
```diff
@@ -300,3 +301,166 @@ def _weightedrankedtau(ordered[:] x, ordered[:] y, intp_t[:] rank, weigher, bool
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+cdef _rank_distance_matrix(ndarray distx):
```
https://github.com/scipy/scipy/pull/10524/commits/58289bddc083213f48320495ef5178599a6189e6

sampan501 commented on scipy/stats/_stats.pyx:
```diff
@@ -300,3 +301,166 @@ def _weightedrankedtau(ordered[:] x, ordered[:] y, intp_t[:] rank, weigher, bool
+cdef _local_covariance(ndarray distx, ndarray disty, ndarray rank_distx, ndarray rank_disty):
+    # convert float32 numpy array to int, as it will be used as array indices
+    # [0 to n-1]
+    rank_distx = rank_distx.astype(np.long, copy=False) - 1
```
https://github.com/scipy/scipy/pull/10524/commits/58289bddc083213f48320495ef5178599a6189e6

sampan501 commented on scipy/stats/_stats.pyx:
```diff
@@ -300,3 +301,166 @@ def _weightedrankedtau(ordered[:] x, ordered[:] y, intp_t[:] rank, weigher, bool
+    cdef int n = distx_view.shape[0]
+    cdef int nx = np.max(rank_distx) + 1
+    cdef int ny = np.max(rank_disty) + 1
+    cdef ndarray cov_xy = np.zeros((nx, ny), dtype=np.float)
```
https://github.com/scipy/scipy/pull/10524/commits/58289bddc083213f48320495ef5178599a6189e6

sampan501 commented on scipy/stats/_stats.pyx:
```diff
@@ -300,3 +301,166 @@ def _weightedrankedtau(ordered[:] x, ordered[:] y, intp_t[:] rank, weigher, bool
+    rank_distx = rank_distx.astype(np.long, copy=False) - 1
+    rank_disty = rank_disty.astype(np.long, copy=False) - 1
+
+    cdef float64_t[:, :] distx_view = distx
```
https://github.com/scipy/scipy/pull/10524/commits/58289bddc083213f48320495ef5178599a6189e6

sampan501 commented on scipy/stats/_stats.pyx:
```diff
@@ -300,3 +301,166 @@ def _weightedrankedtau(ordered[:] x, ordered[:] y, intp_t[:] rank, weigher, bool
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+cpdef _local_correlations(ndarray distx, ndarray disty, str global_corr='mgc'):
```
https://github.com/scipy/scipy/pull/10524/commits/58289bddc083213f48320495ef5178599a6189e6

sampan501 commented on scipy/stats/stats.py:
```diff
@@ -4259,6 +4265,529 @@ def weightedtau(x, y, rank=True, weigher=None, additive=True):
+
+    # convert arrays of type (n,) to (n, 1)
+    if x.ndim == 1:
+        x.shape = (-1, 1)
```
https://github.com/scipy/scipy/pull/10524/commits/fabfbb13356c806099a19d3ce7ed5448ba2219d0

sampan501 commented on scipy/stats/stats.py:
```diff
@@ -4259,6 +4265,529 @@ def weightedtau(x, y, rank=True, weigher=None, additive=True):
+
+    Notes
+    -----
+
```
https://github.com/scipy/scipy/pull/10524/commits/fabfbb13356c806099a19d3ce7ed5448ba2219d0

sampan501 commented on scipy/stats/stats.py:
```diff
@@ -4259,6 +4265,529 @@ def weightedtau(x, y, rank=True, weigher=None, additive=True):
+    pvalue = 0
+    null_dist = np.zeros(reps)
+
+    # use all cores to create function that parallelizes over number of reps
```
https://github.com/scipy/scipy/pull/10524/commits/fabfbb13356c806099a19d3ce7ed5448ba2219d0

sampan501 commented on scipy/stats/stats.py:
```diff
@@ -4259,6 +4265,529 @@ def weightedtau(x, y, rank=True, weigher=None, additive=True):
+    Parameters
+    ----------
+    x, y : ndarray
+        `x` and `y` have shapes `(n, p)` and `(n, q)`.
```
https://github.com/scipy/scipy/pull/10524/commits/fabfbb13356c806099a19d3ce7ed5448ba2219d0

sampan501 commented on scipy/stats/stats.py:
```diff
@@ -4259,6 +4265,529 @@ def weightedtau(x, y, rank=True, weigher=None, additive=True):
+        return perm_stat
+
+
+def _perm_test(x, y, stat, compute_distance, reps=1000, workers=-1):
```
https://github.com/scipy/scipy/pull/10524/commits/fabfbb13356c806099a19d3ce7ed5448ba2219d0

rgommers replied:
a test with `workers=2` would still be nice, decorated with that `skipif`.

rgommers replied:
@sampan501 there's still this comment, that would be quite good to do.

sampan501 replied:
https://github.com/scipy/scipy/pull/10524/commits/18d6b8d7a451458d95017433a06a4e77cd94f707

sampan501 commented on scipy/stats/stats.py:
```diff
@@ -4259,6 +4265,529 @@ def weightedtau(x, y, rank=True, weigher=None, additive=True):
+    >>> x = np.linspace(-1, 1, num=100)
+    >>> y = x + 0.3 * np.random.random(x.size)
+    >>> stat, pvalue, _ = multiscale_graphcorr(x, y)
+    >>> round(stat, 1), round(pvalue, 1)
```
https://github.com/scipy/scipy/pull/10524/commits/fabfbb13356c806099a19d3ce7ed5448ba2219d0

sampan501 commented on scipy/stats/stats.py:
```diff
@@ -4259,6 +4265,529 @@ def weightedtau(x, y, rank=True, weigher=None, additive=True):
+
+    Alternatively,
+
+    >>> from scipy.stats import multiscale_graphcorr
```
https://github.com/scipy/scipy/pull/10524/commits/fabfbb13356c806099a19d3ce7ed5448ba2219d0

sampan501 commented on scipy/stats/stats.py:
```diff
@@ -4259,6 +4265,529 @@ def weightedtau(x, y, rank=True, weigher=None, additive=True):
+        `multiprocessing.Pool <multiprocessing>`). Supply `-1` to use all cores
+        available to the Process. Alternatively supply a map-like callable,
+        such as `multiprocessing.Pool.map` for evaluating the population in
+        parallel. This evaluation is carried out as `workers(func, iterable)`.
```
https://github.com/scipy/scipy/pull/10524/commits/fabfbb13356c806099a19d3ce7ed5448ba2219d0

sampan501 commented on scipy/stats/stats.py:
```diff
@@ -4259,6 +4265,529 @@ def weightedtau(x, y, rank=True, weigher=None, additive=True):
+
+
+def multiscale_graphcorr(x, y, compute_distance=_euclidean_dist, reps=1000,
+                         workers=-1, is_twosamp=False):
```
https://github.com/scipy/scipy/pull/10524/commits/fabfbb13356c806099a19d3ce7ed5448ba2219d0

sampan501 commented:
@rgommers I think that I addressed all of your comments. Let me know if you think I need to change anything else

rgommers commented:
Thanks @sampan501. The one test failure is real:
```
scipy.stats.multiscale_graphcorr
--------------------------------
File "build/testenv/lib/python3.6/site-packages/scipy/stats/stats.py", line 4612, in multiscale_graphcorr
Failed example:
    mgc.stat, mgc.pvalue
Expected:
    (0.033258146255703246, 0.001)
Got:
    (0.033258146255703246, 0.014)
```
looks like `mgc.stat` should be rounded to 3 digits as well

rgommers commented on scipy/stats/stats.py:
```diff
@@ -4259,6 +4265,537 @@ def weightedtau(x, y, rank=True, weigher=None, additive=True):
+    m, n = stat_mgc_map.shape
+
+    # 0.02 is simply an empirical threshold, this can be set to 0.01 or 0.05
+    # with varying levels of performance. Threshold is based on a beta
```
follow-up question: if it affects the performance, should it be a parameter? or is it hard to tune and is 0.02 a reasonable compromise for most inputs?

sampan501 replied:
It is the kind of parameter that is hard to tune. 0.02 works well for most of the simulations that we tested.

rgommers replied:
Okay, then let's leave it as is, thanks.

rgommers commented:
Thanks for the quick updates. Two things left to address.

sampan501 commented:
The azure pipelines are showing me this error.
```
Failures
 - mingw (exited -1) - Error while running 'C:\ProgramData\chocolatey\lib\mingw\tools\chocolateyinstall.ps1'.
 See log for details.
##[error]PowerShell exited with code '1'.
```
Is this likely due to a problem at my end?

tylerjereddy commented:
> Is this likely due to a problem at my end?

It looks like a hash mismatch on the mingw install, so perhaps not. I do see that this feature branch is about 800 commits behind master though, so that may not help.

rgommers commented:
I've seen more of those recently, must be unrelated.

rgommers commented:
Changes all look good, my only comment that isn't yet addressed is adding a test with `workers=2` in https://github.com/scipy/scipy/pull/10524#discussion_r346635542 (note the `skipif` to add to it).

I'll look again in the morning, hopefully we can hit the green button then!

sampan501 commented:
@rgommers Everything seems to work correctly with the new test now. The Travis log error for the one failure is not my own:
```
build/testenv/lib/python3.6/site-packages/scipy/stats/tests/test_mstats_basic.py 
[gw2] linux -- Python 3.6.7 /home/travis/virtualenv/python3.6.7/bin/python
worker 'gw2' crashed while running 'build/testenv/lib/python3.6/site-packages/scipy/stats/tests/test_mstats_basic.py::TestCompareWithStats::test_kendalltau'
```

rgommers commented:
All green, in it goes. Thanks @sampan501! And thanks for everyone else who reviewed or tested.

rgommers commented:
Just in time for 1.4.0:)
